src/esplog/mod.rs

```diff
@@ -0,0 +1,220 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use std::io::*;
+use std::str::from_utf8;
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc,
+};
+
+use byteorder::{LittleEndian, ReadBytesExt};
+
+use crate::esplog::mini_decompressor::{decompress_block, Quat, State};
+use crate::tags_impl::*;
+use crate::*;
+
+mod mini_decompressor;
+
+#[derive(Default)]
+pub struct EspLog {
+    pub model: Option<String>,
+}
+
+impl EspLog {
+    pub fn possible_extensions() -> Vec<&'static str> {
+        vec!["bin"]
+    }
+
+    pub fn detect<P: AsRef<std::path::Path>>(buffer: &[u8], _filepath: P) -> Option<Self> {
+        if buffer.len() > 7 && buffer[0..7] == [0x45, 0x73, 0x70, 0x4c, 0x6f, 0x67, 0x30] {
+            return Some(Self { model: None });
+        }
+        None
+    }
+
+    pub fn parse<T: Read + Seek, F: Fn(f64)>(
+        &mut self,
+        stream: &mut T,
+        size: usize,
+        progress_cb: F,
+        cancel_flag: Arc<AtomicBool>,
+    ) -> Result<Vec<SampleInfo>> {
+        let mut buf = vec![0u8; 8000];
+
+        let mut gyro = Vec::new();
+        let mut accel = Vec::new();
+
+        // skip header
+        stream.read_exact(&mut buf[..7])?;
+
+        let mut accel_blk_size = 0;
+        let mut accel_range = 0i32;
+        let mut state = State::new();
+        let mut tmp_quats = vec![];
+        let mut quats: Vec<Quat> = vec![];
+        let mut accels: Vec<[i16; 3]> = vec![];
+        let mut prev_quat = Quat::default();
+        let mut cur_time = 0.0;
+        let mut orientation = "xyz".to_string();
+        let mut res = || {
+            loop {
+                if cancel_flag.load(Ordering::Relaxed) {
+                    break;
+                }
+                match stream.read_u8()? {
+                    0x01 => {
+                        // gyro setup
+                        if stream.read_u8()? != 0x01 {
+                            return Err(Error::new(ErrorKind::Other, "Unsupported algo revision"));
+                        }
+                        let blk_size = stream.read_u16::<LittleEndian>()?;
+                        tmp_quats.resize(blk_size as usize, Quat::default());
+                    }
+                    0x02 => {
+                        // delta time
+                        let dt = stream.read_u32::<LittleEndian>()?;
+
+                        // generate timestamps for gyro
+                        let scale =
+                            quats.len() as f64 / dt as f64 * 1e6 * 180.0 / std::f64::consts::PI;
+                        for (i, &q) in quats.iter().enumerate() {
+                            let rv = (q.conj() * prev_quat).to_rvec();

```

Comment:
Since it decompresses to quats, you can also populate the `GroupId::Quaternions` directly - is there any benefit to it? Does converting it to rvec lose any information? If it's equivalent, then there's no point of adding Quaternions